lib/dao.js

```diff
@@ -624,7 +624,184 @@ DataAccessObject.upsert = function(data, options, cb) {
   }
   return cb.promise;
 };
+/**
+ * Performs upsert based on the instance that matches the where clause
+ . This is a PUT operation and based on
+ * non-primary key. The filter takes one key value pair and updates single matching instance.*
+ */
+DataAccessObject.patchOrCreateWithWhere =
+  DataAccessObject.upsertWithWhere = function(where, data, options, cb) {
+    var connectionPromise = stillConnecting(this.getDataSource(), this, arguments);
+    if (connectionPromise) { return connectionPromise; }
+    assert(arguments.length >= 1, 'At least one argument is required');
+    if (data === undefined && options === undefined && cb === undefined && arguments.length === 1) {
+      data = where;
+      where = {};
+    } else if (options === undefined && cb === undefined) {
+      if (typeof data === 'function') {
```

Comment:
+1 I agree and this issue is caught/mentioned in the previous round of reviews as well; please see [this comment](https://github.com/strongloop/loopback-datasource-juggler/pull/1001#discussion_r70968475) and [this comment](https://github.com/strongloop/loopback-datasource-juggler/pull/1001#discussion_r70968610)